subsys/fb/Kconfig

```diff
@@ -31,6 +31,26 @@ config CHARACTER_FRAMEBUFFER_SHELL_DRIVER_NAME
 	help
 	  Character Framebuffer Display Driver Name
 
+config CHARACTER_FRAMEBUFFER_USE_OFFSCREEN_BUFFER
+	bool "Use offscreen buffer"
+	default y
+	help
+	  Allocate and use an offscreen buffer for rendering.
+	  If this option is disabled, some functions that depend on
+	  offscreen buffers are unsupported.
```

Comment:
I think it is a confusing name. What is offscreen buffer? It still uses both smaller buffer (provided by cfb) and obviously display controller buffer.
Better not to introduce this option at all, just make everything work with a configurable smaller buffer, otherwise there is a risk that this problematic situation will never change.

Reply:
> Better not to introduce this option at all, just make everything work with a configurable smaller buffer, otherwise there is a risk that this problematic situation will never change.

In reality, it is difficult(or cannot).

Because many monochrome displays, such as the SSD1306, cannot draw by individual dots.


Without the offscreen buffer, drawing freedom is greatly restricted.
For example,...

It is generally possible to draw simple characters, but it is not possible to draw graphics.
Performing the drawing is in units of 1x8 (VTILED). It will lost the drawing contents by the unit.

for example,
cfb_draw_line( {0, 2}, {10, 2} );
cfb_draw_line( {0, 6}, {10, 6} );

and try to draw two horizontal lines but in this case,
The first drawn line is lost since it draws on the same 8-dot tile.

The requirement is also unfamiliar if you support drawing fonts that are not multiples of 8 dots in the vertical direction.
The user needs to be conscious of the area where the characters are drawn and the boundary of 8 dot units.
When drawing with a vertical 9-dot font, the user must consider the 8-dot boundary and draw the subsequent text from the 16th dot.
If the user does not care about the border, such as drawing from the vertical 10-dot line,
will lose the bottom 1-dot of the font. 

It also restricts the ability to express graphics.
For example, you can't use a technique that draws pseudo-bold by shifting the characters slightly and drawing them twice.

If it can obtain the current drawing content  from the display, 
It can handle a read-modify-write way, but reading from the display is usually unsupported.

Therefore, I think it is not desirable only to support drawing with a small-size transfer buffer. think it is not desirable only to support drawing with a small-size transfer buffer.

Reply:
I'm also confused about the goal of this mode. As I understand it, when `CONFIG_CHARACTER_FRAMEBUFFER_USE_OFFSCREEN_BUFFER` is set, a partial framebuffer is allocated versus a screen sized one. This is similar to LVGL, where software allocated framebuffers do not have to be as large as the display.

Is the idea that a user can set `CONFIG_CHARACTER_FRAMEBUFFER_USE_OFFSCREEN_BUFFER` when they do not need/want to allocate a buffer that is the size of their output display?

Reply:
Yes, it is. It works with even extremely tiny buffers when CONFIG_CHARACTER_FRAMEBUFFER_USE_OFFSCREEN_BUFFER is disabled.

The offscreen buffer, which is the same size as the display, remains left for supporting graphics rendering.

Reply:
@jfischer-no @danieldegrasse 

As a suggestion, I created PR (#61533) to introduce the u8g2 library to improve the graphic rendering function.
u8g2 can offscreen-buffer-based graphics rendering. It decrease the necessity to implement graphics function in CFB.
If so, we can remove the buffer from CFB and optimize for minimal character rendering.
(In other words, it requires dropping the graphics rendering feature to remove the buffer.)

Reply:
> > Better not to introduce this option at all, just make everything work with a configurable smaller buffer, otherwise there is a risk that this problematic situation will never change.
> 
> In reality, it is difficult(or cannot).
> 
> Because many monochrome displays, such as the SSD1306, cannot draw by individual dots.
> 
> Without the offscreen buffer, drawing freedom is greatly restricted. For example,...
> 
> It is generally possible to draw simple characters, but it is not possible to draw graphics. Performing the drawing is in units of 1x8 (VTILED). It will lost the drawing contents by the unit.
> 
> for example, cfb_draw_line( {0, 2}, {10, 2} ); cfb_draw_line( {0, 6}, {10, 6} );

Besides limited displays like monochrome, library would be rewriting the same array many times if there are more than one graphic objects, so the problem is there anyway. To solve this, cfb (which should be renamed to something else, e.g. zgl) can group the objects in the same area and draw that area only once. So cfb_draw_line() would not really draw a line on display but register this object and perhaps place it in right order.

Reply:
Hmm, it is theoretically possible to do it.
But I wonder if it is a good way or not.
I think it does not work with implementing a simple rendering object list.
It needs to calculate what object needs to be rendered in the position.
It is a hit detection.
It is like reducing the dimension from the 3D engine by projection.
I feel it is too complex for tiny systems.

This is not culling hidden surfaces problem,
but the buffer will take a significant role, like as z-buffer in 3d rendering, I think.